http-clients/src/main/java/com/palantir/remoting3/clients/UserAgent.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting3.clients;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.base.Joiner;
+import com.google.common.collect.Lists;
+import java.util.List;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import org.immutables.value.Value;
+
+/**
+ * Constructs, validates, and formats a canonical User-Agent header. Because the http header spec
+ * (https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2) requires headers to be joined on commas, individual
+ * {@link Agent} header strings must never contain commas.
+ */
+@Value.Immutable
+@ImmutablesStyle
+public interface UserAgent {
+    List<Agent> agents();
+
+    @Value.Lazy
+    default String headerFormat() {
+        return Joiner.on(", ").join(Lists.transform(agents(), Agent::headerFormat));
+    }
+
+    static UserAgent of(String serviceName, String instanceId, String version) {
+        Agent agent = ImmutableAgent.builder()
+                .serviceName(serviceName)
+                .instanceId(instanceId)
+                .version(version)
+                .build();
+        return ImmutableUserAgent.builder()
+                .addAgents(agent)
+                .build();
+    }
+
+    static UserAgent of(String serviceName, String version) {
+        Agent agent = ImmutableAgent.builder()
+                .serviceName(serviceName)
+                .version(version)
+                .build();
+        return ImmutableUserAgent.builder()
+                .addAgents(agent)
+                .build();
+    }
+
+    @Value.Default
+    default UserAgent append(String serviceName, String version) {
```

Comment:
Wonder if this should be a static utility so the use is more clear.  Append in this case could suggest that this could mutate the original.  Could lead to bugs in consumers.

Reply:
so something like: `UserAgent.concat(originalAgent, String serviceName, String version)`

Reply:
I'm personally in favor of this given that the `UserAgent` is immutable. Willing to change though if another person finds `UserAgent.append(userAgent, service, version)` more clear than `userAgent.append(service, version)`.

Reply:
Personally i wouldn't even have this method. I do think Matt has a point around appending having a mutability sound to it.

Reply:
done